From 6b600fe273068ff3dc7133c61e295890c2577e8b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernhard=20=C3=9Cbelacker?= <bernhardu@mailbox.org>
Date: Tue, 10 Dec 2024 01:06:53 +0100
Subject: [PATCH 41/86] user32/tests: wine_dbgstr_w - wine_dbgstr_wn

=================================================================
==1048==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffffe1fe4d0 at pc 0x000140160cb1 bp 0x7ffffe1fc0a0 sp 0x7ffffe1fc0e8
READ of size 2 at 0x7ffffe1fe4d0 thread T0
0414:fixme:file:server_get_file_info Unsupported info class e
    #0 0x000140160cb0 in wine_dbgstr_wn /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/debug.h:263:30
    #1 0x0001401155dc in wine_dbgstr_w /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/debug.h:306:12
    #2 0x0001401155dc in test_key_names /home/bernhard/data/entwicklung/2024/wine\wine/dlls/user32/tests/input.c:3838:49
    #3 0x0001401155dc in func_input /home/bernhard/data/entwicklung/2024/wine\wine/dlls/user32/tests/input.c:6180:5
    #4 0x00014044f751 in run_test /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:765:5
    #5 0x00014044f751 in main /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:884:12
    #6 0x0001404516cf in mainCRTStartup /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msvcrt/crt_main.c:58:11
    #7 0x6ffffc1c4808 in BaseThreadInitThunk /home/bernhard/data/entwicklung/2024/wine/wine/dlls/kernel32\thread.c:61:5
    #8 0x6ffffc39fa1a in RtlUserThreadStart (C:\windows\system32\ntdll.dll+0x17000fa1a)

Address 0x7ffffe1fe4d0 is located in stack of thread T0 at offset 8336 in frame
    #0 0x00014010ad1f in func_input /home/bernhard/data/entwicklung/2024/wine\wine/dlls/user32/tests/input.c:6147

  This frame has 68 object(s):
    [32, 48) 'rect.i1566' (line 5786)
    [64, 80) 'clip_rect.i1567' (line 5786)
    [96, 112) 'virtual_rect.i1568' (line 5786)
    [128, 200) 'cls.i' (line 5156)
    [240, 320) 'inputs.i' (line 5157)
    [352, 368) 'rc.i' (line 5159)
    [384, 432) 'msg.i1478' (line 5160)
    [464, 976) 'devices.i' (line 1913)
    [1040, 1044) 'devcount.i' (line 1914)
    [1056, 1060) 'odevcount.i' (line 1914)
    [1072, 1104) 'info.i' (line 1949)
    [1136, 1140) 'size.i' (line 1950)
    [1152, 1408) 'name.i' (line 1986)
    [1472, 1600) 'nameA.i' (line 1987)
    [1632, 1636) 'sz.i' (line 1988)
    [1648, 1680) 'info143.i' (line 1989)
    [1712, 1720) 'preparsed.i' (line 2051)
    [1744, 1768) 'in.i1197' (line 1604)
    [1808, 6608) 'out.i1198' (line 1605)
    [6864, 6872) 'point.i1199' (line 1606)
    [6896, 6936) 'input.i' (line 1607)
    [6976, 7008) 'params.i1014' (line 2998)
    [7040, 7064) 'process_info.i' (line 2999)
    [7104, 7120) 'raw_devices.i1015' (line 3000)
    [7136, 7240) 'startup_info.i' (line 3001)
    [7280, 7288) 'pt.i1016' (line 3004)
    [7312, 7320) 'newpt.i' (line 3004)
    [7344, 7604) 'path.i' (line 3007)
    [7680, 7682) 'wchr.i' (line 5071)
    [7696, 7697) 'oem_char.i' (line 5072)
    [7712, 7736) 'params.i' (line 4959)
    [7776, 8032) 'keystate.i' (line 4962)
    [8096, 8144) 'msg.i853' (line 4965)
    [8176, 8216) 'buffer.i' (line 3804)
    [8256, 8336) 'bufferW.i' (line 3805) <== Memory access at offset 8336 overflows this variable
    [8368, 8386) 'klid.i' (line 3476)
    [8432, 8450) 'tmpklid.i' (line 3476)
    [8496, 9016) 'layout_path.i' (line 3476)
    [9152, 9162) 'value.i' (line 3476)
    [9184, 9188) 'value_size.i' (line 3478)
    [9200, 9204) 'klid_size.i' (line 3478)
    [9216, 9220) 'type.i' (line 3478)
    [9232, 9240) 'hkey.i' (line 3480)
    [9264, 9296) 'wstr.i' (line 3383)
    [9328, 9344) 'str.i' (line 3384)
    [9360, 9362) 'character.i' (line 3385)
    [9376, 9632) 'state.i409' (line 3386)
    [9696, 9704) 'wStr.i' (line 3309)
    [9728, 9984) 'state.i' (line 3310)
    [10048, 10304) 'buff.i' (line 1446)
    [10368, 10384) 'rect.i247' (line 5730)
    [10400, 10416) 'clip_rect.i248' (line 5730)
    [10432, 10448) 'virtual_rect.i249' (line 5730)
    [10464, 10480) 'rect.i218' (line 5677)
    [10496, 10512) 'clip_rect.i' (line 5677)
    [10528, 10544) 'virtual_rect.i' (line 5677)
    [10560, 10568) 'thread.i' (line 5679)
    [10592, 10608) 'rect.i' (line 5634)
    [10624, 10648) 'in.i' (line 1828)
    [10688, 12224) 'out.i' (line 1829)
    [12352, 13888) 'out2.i' (line 1829)
    [14016, 14024) 'point.i' (line 1830)
    [14048, 14064) 'raw_devices.i' (line 2764)
    [14080, 14088) 'done.i' (line 2765)
    [14112, 14120) 'pt.i' (line 2767)
    [14144, 14192) 'msg.i' (line 2769)
    [14224, 14232) 'argv' (line 6148)
    [14256, 14264) 'pos' (line 6150)
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp, SEH and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/debug.h:263:30 in wine_dbgstr_wn
Shadow bytes around the buggy address:
  0x7ffffe1fe200: f2 f2 f2 f2 f2 f2 f2 f2 f8 f2 f8 f2 f8 f8 f8 f2
  0x7ffffe1fe280: f2 f2 f2 f2 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8
  0x7ffffe1fe300: f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8
  0x7ffffe1fe380: f8 f8 f8 f8 f2 f2 f2 f2 f2 f2 f2 f2 f8 f8 f8 f8
  0x7ffffe1fe400: f8 f8 f2 f2 f2 f2 00 00 00 00 00 f2 f2 f2 f2 f2
=>0x7ffffe1fe480: 00 00 00 00 00 00 00 00 00 00[f2]f2 f2 f2 f8 f8
  0x7ffffe1fe500: f8 f2 f2 f2 f2 f2 f8 f8 f8 f2 f2 f2 f2 f2 f8 f8
  0x7ffffe1fe580: f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8
  0x7ffffe1fe600: f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8
  0x7ffffe1fe680: f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8
  0x7ffffe1fe700: f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f8 f2
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==1048==ABORTING
0414:fixme:kernelbase:AppPolicyGetProcessTerminationMethod FFFFFFFFFFFFFFFA, 00007FFFFEA8FE80
make: *** [Makefile:387435: dlls/user32/tests/x86_64-windows/input.ok] Fehler 1
---
 dlls/user32/tests/input.c | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/dlls/user32/tests/input.c b/dlls/user32/tests/input.c
index 9196e67c15b..0bffb40d718 100644
--- a/dlls/user32/tests/input.c
+++ b/dlls/user32/tests/input.c
@@ -3824,19 +3824,19 @@ static void test_key_names(void)
 
     memset( bufferW, 0xcc, sizeof(bufferW) );
     ret = GetKeyNameTextW( lparam, bufferW, ARRAY_SIZE(bufferW));
-    ok( ret > 0, "wrong len %u for %s\n", ret, wine_dbgstr_w(bufferW) );
-    ok( ret == lstrlenW(bufferW), "wrong len %u for %s\n", ret, wine_dbgstr_w(bufferW) );
+    ok( ret > 0, "wrong len %u for %s\n", ret, wine_dbgstr_wn(bufferW, ARRAY_SIZE(bufferW)) );
+    ok( ret == lstrlenW(bufferW), "wrong len %u for %s\n", ret, wine_dbgstr_wn(bufferW, ARRAY_SIZE(bufferW)) );
 
     memset( bufferW, 0xcc, sizeof(bufferW) );
     prev = ret;
     ret = GetKeyNameTextW( lparam, bufferW, prev );
-    ok( ret == prev - 1, "wrong len %u for %s\n", ret, wine_dbgstr_w(bufferW) );
-    ok( ret == lstrlenW(bufferW), "wrong len %u for %s\n", ret, wine_dbgstr_w(bufferW) );
+    ok( ret == prev - 1, "wrong len %u for %s\n", ret, wine_dbgstr_wn(bufferW, ARRAY_SIZE(bufferW)) );
+    ok( ret == lstrlenW(bufferW), "wrong len %u for %s\n", ret, wine_dbgstr_wn(bufferW, ARRAY_SIZE(bufferW)) );
 
     memset( bufferW, 0xcc, sizeof(bufferW) );
     ret = GetKeyNameTextW( lparam, bufferW, 0 );
-    ok( ret == 0, "wrong len %u for %s\n", ret, wine_dbgstr_w(bufferW) );
-    ok( bufferW[0] == 0xcccc, "wrong string %s\n", wine_dbgstr_w(bufferW) );
+    ok( ret == 0, "wrong len %u for %s\n", ret, wine_dbgstr_wn(bufferW, ARRAY_SIZE(bufferW)) );
+    ok( bufferW[0] == 0xcccc, "wrong string %s\n", wine_dbgstr_wn(bufferW, ARRAY_SIZE(bufferW)) );
 }
 
 static void simulate_click(BOOL left, int x, int y)
-- 
2.47.1

